src/Npgsql/Internal/NpgsqlConnector.FrontendMessages.cs

```diff
@@ -227,7 +227,7 @@ internal async Task WriteParse(string sql, byte[] asciiName, List<NpgsqlParamete
 
         writeBuffer.WriteUInt16((ushort)parameters.Count);
 
-        var writer = writeBuffer.GetWriter(DatabaseInfo, async ? FlushMode.NonBlocking : FlushMode.Blocking);
+        var writer = parameters.Count > 0 ? writeBuffer.GetWriter(DatabaseInfo, async ? FlushMode.NonBlocking : FlushMode.Blocking) : null!;
```

Comment:
Maybe just do an if and put all of try/catch with getting a writer there?

Reply:
Sure

Reply:
I really dislike how we have so many loops over a List... IMO we should just move our NpgsqlParameterCollection to use an array, manually growing it.

Reply:
> I really dislike how we have so many loops over a List... IMO we should just move our NpgsqlParameterCollection to use an array, manually growing it.

We can experiment with this but in a separate pr.